File: Code Generation/responses/cpp/gemini/medium/longest-palindromic-subsequence-after-at-most-k-operations.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cstring>

using namespace std;

class Solution {
    int dp[201][201][201];

    int getDist(char a, char b) {
        int x = a - 'a';
        int y = b - 'a';
        int diff = abs(x - y);
        return min(diff, 26 - diff);
    }

public:
    int longestPalindromicSubsequence(string s, int k) {
        int n = s.length();
        // Initialize DP table with -1
        memset(dp, -1, sizeof(dp));
        return solve(0, n - 1, k, s);
    }

    int solve(int i, int j, int k, const string& s) {
        if (i > j) return 0;
        if (i == j) return 1;
        if (dp[i][j][k] != -1) return dp[i][j][k];

        // Option 1: Skip s[i]
        int res = solve(i + 1, j, k, s);
        
        // Option 2: Skip s[j]
        res = max(res, solve(i, j - 1, k, s));

        // Option 3: Match s[i] and s[j] by changing them to the same character
        // We want to find the minimum cost to make s[i] == s[j].
        // This is simply the circular distance between the two characters.
        int cost = getDist(s[i], s[j]);
        if (k >= cost) {
            res = max(res, 2 + solve(i + 1, j - 1, k - cost, s));
        }

        return dp[i][j][k] = res;
    }
};

File: Code Generation/responses/python/gemini/medium/longest-palindromic-subsequence-after-at-most-k-operations.py
import sys
from functools import cache

sys.setrecursionlimit(2000)

class Solution:
    def longestPalindromicSubsequence(self, s: str, k: int) -> int:
        @cache
        def dp(i: int, j: int, rem: int) -> int:
            if i > j:
                return 0
            if i == j:
                return 1
            
            res = dp(i + 1, j, rem)
            res2 = dp(i, j - 1, rem)
            if res2 > res:
                res = res2
            
            cost = abs(ord(s[i]) - ord(s[j]))
            cost = cost if cost <= 13 else 26 - cost
            
            if cost <= rem:
                res3 = 2 + dp(i + 1, j - 1, rem - cost)
                if res3 > res:
                    res = res3
                    
            return res

        return dp(0, len(s) - 1, k)